akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
     system: ExtendedActorSystem,
     phases: Map[String, CoordinatedShutdown.Phase])
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
+
+  private trait PhaseDefinition {
+    def size: Int
+    def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done]
+  }
+
+  private object tasks {
+    private val registeredPhases = new ConcurrentHashMap[String, StrictPhaseDefinition]()
+    private trait TaskDefinition extends Cancellable {
+      private[tasks] def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done]
+    }
+    private object TaskDefinition {
+      def create(phaseName: String, task: () => Future[Done], name: String): TaskDefinition = new TaskDefinition {
+        // This is a vanilla class instead of a case class to avoid default implementations of .hashCode() and .equals
+        // Different code paths could register the same task under the same name multiple times; in this case we want to
+        // run that task as many times as it was registered (minus the number of times those were cancelled), so they must
+        // be distinct in a Set[TaskDefinition].
+
+        private val p = Promise[Done]()
```

Comment:
Hmm, couldn't this just as well be a volatile `completedOrCancelled` boolean, we want to track the intent to run or cancel it, not that it is done, triggering run, and then cancel before the task has completed will still look like a successful cancel right now.

Reply:
Sorry, not `volatile` since then toggling in cancel and knowing that change was the one would probably need an `AtomicBoolean`.

Reply:
First, dang you to heck for sending me down a rabbit hole only to discover that I've been using Promises in a subtly incorrect way for years. (No, seriously, good catch.)

Second, I don't think we're concerned about cancellation triggered after the phase begins. Just as a CS task that adds another task to a later phase must be honored (unless it makes the request after its phase's timeout has expired) but a task which tries to modify its own phase or an earlier phase need not be honored, a cancellation after a task's phase has begun need not be honored. In any case, I think the revised version avoids the behavior you envision.

Reply:
Yup, looks good now!